### ISSUE TITLE
Add elastic json encoding to kafka log exporter

### ISSUE THREAD

**Issue by goto1134** [open-telemetry/opentelemetry-collector-contrib#7416]
**Is your feature request related to a problem? Please describe.**
In my infrastructure we use kafka to collect logs in elastic json format.

**Describe the solution you'd like**
I would like to configure elastic marshaller for kafka log exporter and send logs in elastic format.

**Comment by jkowall** [comment]
@goto1134 Can you explain the flow? I assume you log to Kafka, and then you want the collector to take the logs from Kafka and write them to ElasticSearch?

**Comment by jkowall** [comment]
You can already do this with the Kafka exporter found here: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/kafkaexporter

As far as what "elastic json format" is I am not sure what you mean. The collector will work with JSON in OLTP format, but it doesn't know the "elastic json format" specifically.

**Comment by MovieStoreGuy** [comment]
A reference to expected format would help here @goto1134 and I happy to consider it, but I will warn you that adding json exported types to the collector requires a lot more resources compared to consuming protobuf; hence the otlp-json encoding is not recommended for production use.

**Comment by cyrille-leclerc** [comment]
Hello @goto1134 . Elastic v8.0 offers native support for OpenTelemetry logs on the Elastic OTLP intake. 

**Would this solve your problem?**

The architecture would be:
```
                                                                                    ┌────────────────────────────────────────┐
                                                                                    │                ELASTIC                 │
                                                                                    │                                        │
    ┌────────────────────┐ OTLP  ┌──────────────┐  ┌─────┐  ┌──────────────┐ OTLP   │ ┌─────────────────┐    ┌─────────────┐ │
    │ Otel Logs Producer ├──────►│Otel Collector├─►│Kafka├─►│Otel Collector├────────┤►│Elstic APM Server│    │Elasticsearch│ │
    └────────────────────┘       └──────────────┘  └─────┘  └──────────────┘        │ └─────────────────┘    └─────────────┘ │
                                                                                    │                                        │
                                                                                    └────────────────────────────────────────┘
```

More details on the Elastic native support for OpenTelemetry on https://www.elastic.co/guide/en/apm/guide/current/open-telemetry.html

**Comment by goto1134** [comment]
Hello!

The format I am looking for is a json
```json
{
"@timestamp" : ""
"@version": ""
"message": ""
}
```

with the ability to add custom fields at any level of this object. We use [logback-logstash-encoder](https://github.com/logfellow/logstash-logback-encoder) to use this json log format.

**Comment by goto1134** [comment]
@cyrille-leclerc , as for the possible integration, we are using a third-party proprietary logging system, that supports only the mentioned logstash format, and the only way to send the logs there is via Kafka.

**Comment by goto1134** [comment]
Yes, of course, but the provider does not plan to implement this.

**Comment by MovieStoreGuy** [comment]
Hey @goto1134 ,

Would a custom worker that just translates from otlp to the elastic json format from one topic to another? 
(There is a risk that it would at least double the number of bytes stored within the kafka cluster).

**Comment by MovieStoreGuy** [comment]
Alright, here is my thoughts on what has been said so far.

A vendor specific encoding goes against the spirit of open telemetry trying to develop an ecosystem of adopting non vendor specific standard; any currently open components should only adopt open standards that multiple vendors / platforms can consume from. 

Having said this, extending the open to be optimised for the vendor specific platform is acceptable (re: aws implementation of open source products) done in making the the non vendor specific component, as a vendor specific component.
Meaning, that this should be created as `elastickafkaexporter` as a community funded project that is heavily optimised to work with your exact use case. 

My rationale of my thoughts (and my personal thoughts, not a reflection of those in the open telemetry project) is that the existing encodings are well established as open standards making it easier as a community to help make improvements, where vendor specific encodings (in this scenario) is harder for that community (specifically me as one of the code owners) to support. I unfortunately don't use any elastic products (personally or for work) so my ability to help personally is extremely limited since I can't go through my usual approach of validating changes, and I would be relying on someone like you @goto1134 to validate the changes for me. (which is a really shitty development loop, and I for one wouldn't wish that on anyone :\ )

To recap my thoughts here, yes I agree that there is a need for vendor specific encodings within open source exporters (like kafka, kinesis) that basically just byte forwarders, however, they should be bundled in a vendor specific exporter of the open source byte forwarded (for a lake of a better term).

**Comment by jkowall** [comment]
> Yes, of course, but the provider does not plan to implement this.

@goto1134  Can we call out the provider so others can apply pressure for them to evolve their thinking? I am happy to speak to them and explain why this important.

**Comment by jpkrohling** [comment]
>, yes I agree that there is a need for vendor specific encodings within open source exporters

Just throwing ideas around: how about having someting similar to the auth extensions, where encoders are extensions and this exporter would then hold a reference to the encoder to use? This way, we can ship with open standards, while people can build their own distributions embedding their vendor-specific ~~lock-in~~ encoder if they want.

**Comment by MovieStoreGuy** [comment]
> where encoders are extensions and this exporter would then hold a reference to the encoder to use?

This would be a cool idea, also if we can add in compression as well since that is kinda tossed around a lot in each repo.
I can at least speak for the kafka, kinesis exporters, were we'd really only need to something like:

```golang
type Encoder interface {
    Encode[P pdata.Typeish](in P)(out []byte, error)
}
```
(Don't mind me practicing my go 1.18, we'll get there.)

It could just exist as a package that those exporter who can emit any encoding, can use. I don't even think it needs to be an extension really, it could just be an encodings package.

**Comment by jpkrohling** [comment]
> I don't even think it needs to be an extension really, it could just be an encodings package.

Extensions can be queried when the component is being built, as they are part of the `component.Host`. Otherwise, we'd need a global registry, which doesn't work well in cases where the collector can be embedded into other binaries.

**Comment by MovieStoreGuy** [comment]
That would be a fun project to add in the collector, and remove potentially a lot of duplicated code I imagine.

Worth separating out into a different issue @jpkrohling ?

**Comment by jpkrohling** [comment]
If you want to work on that, do prepare a design doc and circulate it among the approvers/maintainers. It's not the first time we thought about this, but I think we just didn't have enough engineering power to pursue it.
